File: main/low/src/midi.rs
#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]

use super::bindings::root::reaper_midi::*;
use super::bindings::root::{midi_Input, MIDI_event_t, MIDI_eventlist};
use crate::bindings::root::midi_Output;
use std::os::raw::c_int;
use std::ptr::NonNull;

impl midi_Input {
    pub fn GetReadBuf(&mut self) -> *mut MIDI_eventlist {
        unsafe { midi_Input_GetReadBuf(self as _) }
    }
}

impl MIDI_eventlist {
    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn EnumItems(&mut self, bpos: *mut c_int) -> *mut MIDI_event_t {
        MIDI_eventlist_EnumItems(self as _, bpos)
    }

    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn AddItem(&mut self, evt: *mut MIDI_event_t) {
        MIDI_eventlist_AddItem(self as _, evt);
    }

    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn DeleteItem(&mut self, bpos: c_int) {
        MIDI_eventlist_DeleteItem(self as _, bpos);
    }

    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn GetSize(&mut self) -> c_int {
        MIDI_eventlist_GetSize(self as _)
    }

    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn Empty(&mut self) {
        MIDI_eventlist_Empty(self as _);
    }
}

impl midi_Output {
    /// # Safety
    ///
    /// REAPER can crash if you pass an invalid pointer.
    pub unsafe fn SendMsg(&self, msg: *mut MIDI_event_t, frame_offset: c_int) {
        midi_Output_SendMsg(self as *const _ as _, msg, frame_offset);
    }

    pub fn Send(&self, status: u8, d1: u8, d2: u8, frame_offset: c_int) {
        unsafe { midi_Output_Send(self as *const _ as *mut _, status, d1, d2, frame_offset) };
    }
}

/// Destroys a C++ `midi_Output` object.
///
/// # Safety
///
/// REAPER can crash if you pass an invalid pointer because C++ will attempt to free the wrong
/// location in memory.
pub unsafe fn delete_midi_output(midi_output: NonNull<midi_Output>) {
    crate::bindings::root::reaper_midi::delete_midi_output(midi_output.as_ptr());
}

File: main/low/src/midi.cpp
#include "midi.hpp"

namespace reaper_midi {
  // MIDI_eventlist methods

  MIDI_event_t* MIDI_eventlist_EnumItems(MIDI_eventlist* self, int* bpos) {
    return self->EnumItems(bpos);
  }

  void MIDI_eventlist_AddItem(MIDI_eventlist* self, MIDI_event_t* evt) {
    self->AddItem(evt);
  }

  // midi_Input methods

  MIDI_eventlist* midi_Input_GetReadBuf(midi_Input* self) {
    return self->GetReadBuf();
  }

  // midi_Output methods

  void midi_Output_Send(midi_Output* self, unsigned char status, unsigned char d1, unsigned char d2, int frame_offset) {
    self->Send(status, d1, d2, frame_offset);
  }

  void midi_Output_SendMsg(midi_Output* self, MIDI_event_t* msg, int frame_offset) {
    self->SendMsg(msg, frame_offset);
  }
}